### PR TITLE
[#3033] Add concentration actor method and hooks

### PR DESCRIPTION
**Pull request by krbz999**
Adding the actor method to roll a concentration saving throw, allowing for a DC and other properties to be provided.

Adds two standard hooks.

Making use of `concentration.roll.min` and `.max` for modifying the D20Roll is left as follow-up work since this is [blocked by the roll refactor](https://github.com/foundryvtt/dnd5e/pull/1760#issuecomment-1866519625).

### CODE REVIEW

**Review by arbron — APPROVED** [review]

File: module/documents/actor/actor.mjs [hunk 1]
```diff
@@ -1709,6 +1709,64 @@ export default class Actor5e extends SystemDocumentMixin(Actor) {
...
+    if ( !conc ) throw new Error("You may not make a Concentration Saving Throw with this Actor.");
+
+    const config = CONFIG.DND5E;
+    const modes = CONFIG.Dice.D20Roll.ADV_MODE;
+    const parts = [];
+
+    // Concentration bonus
+    if ( conc.bonus ) {
+      parts.push(conc.bonus);
+    }
```
Comment (Fyorl) [inline module/documents/actor/actor.mjs:1729]:
```suggestion
    if ( conc.bonus ) parts.push(conc.bonus);
```

File: module/documents/actor/actor.mjs [hunk 2]
```diff
@@ -1709,6 +1709,64 @@ export default class Actor5e extends SystemDocumentMixin(Actor) {
...
+
+    const ability = (conc.ability in config.abilities) ? conc.ability : config.defaultAbilities.concentration;
```
Comment (Fyorl) [inline module/documents/actor/actor.mjs:1731]: Do we want to allow for a way to override this via some options passed to this method?

File: module/documents/actor/actor.mjs [hunk 3]
```diff
@@ -1709,6 +1709,64 @@ export default class Actor5e extends SystemDocumentMixin(Actor) {
...
+
+    options = foundry.utils.mergeObject({
+      isConcentration: true,
+      targetValue: 10,
+      advantage: options.advantage || (conc.roll.mode === modes.ADVANTAGE),
+      disadvantage: options.disadvantage || (conc.roll.mode === modes.DISADVANTAGE),
```
Comment (Fyorl) [inline module/documents/actor/actor.mjs:1737]: I might have missed something but I think `mode` is a top level property of `RollConfigField`:

```suggestion
      advantage: options.advantage || (conc.mode === modes.ADVANTAGE),
      disadvantage: options.disadvantage || (conc.mode === modes.DISADVANTAGE),
```

Maybe it should be under `roll`?

File: module/documents/actor/actor.mjs [hunk 4]
```diff
@@ -1709,6 +1709,64 @@ export default class Actor5e extends SystemDocumentMixin(Actor) {
...
+    /**
+     * A hook event that fires before a saving throw to maintain concentration is rolled for an Actor.
+     * @function dnd5e.preRollConcentration
+     * @memberof hookEvents
+     * @param {Actor5e} actor                   Actor for which the saving throw is being rolled.
+     * @param {D20RollConfiguration} options    Configuration data for the pending roll.
+     * @param {string} abilityId                ID of the ability being rolled as defined in `DND5E.abilities`.
+     * @returns {boolean}                       Explicitly return `false` to prevent the save from being performed.
+     */
+    if ( Hooks.call("dnd5e.preRollConcentration", this, options, ability) === false ) return;
```
Comment (Fyorl) [inline module/documents/actor/actor.mjs:1750]: We could consider allowing this hook to modify which ability is used for the concentration save also, but I do not feel strongly.

**Review by Fyorl — CHANGES_REQUESTED** [review]
This is great, thank you for taking care of this work. I have left some comments that need clarification. With regards to overriding the ability used, I do not feel particularly strongly. It's ok if we defer purely to whatever is in the data model at the time of rolling.

File: module/documents/actor/actor.mjs (hunk 3 above)
Comment (krbz999) [reply module/documents/actor/actor.mjs:1737]: We could maybe discuss that on my PR to add the `RollConfig` field to other areas of the actor model, but I'm not opposed. The original idea was `roll` being *modifiers* for the roll, but I suppose that advantage mode very much is that as well.

File: module/documents/actor/actor.mjs (hunk 4 above)
Comment (krbz999) [reply module/documents/actor/actor.mjs:1750]: That's OK I have the strength for both of us.

**Review by Fyorl — APPROVED** [review]

**Pull request merged**